### PR TITLE
feat: improved tree-shaking when importing individual components

### PR DESCRIPTION
**Pull request by tmorehouse**
### Describe the PR

Create simple plugin object factory function, which can be marked as `/*#__PURE__*/` when exporting a plugin (safe for exclusion with no side effects). Rollup and Webpack/Terser will sometimes not exclude exports that are results of function calls, as they are unsure if the function call has any side effects, so whenever we call `Vue.extend(..)`, `pluginFactory(...)`, or `installFactory(...)`, we prefix it with a `/*#__PURE__*/` comment. For better tree shaking and dead code elimination (DCE)  of `dist/bootstrap-vue.esm.js` bundle (if anyone uses it, and is doing selective imports). Based on https://iamakulov.com/notes/polished-webpack/

Also includes a tweak for component and directive top level exports for even better individual component/directive import tree-shaking (until Webpack v5 is released: https://github.com/webpack/webpack/pull/9203, https://github.com/webpack/webpack/issues/8057)

### PR checklist

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] Feature
- [x] Enhancement
- [ ] ARIA accessibility
- [ ] Documentation update
- [x] Other (please describe) code refactor for improved tree shaking

**Does this PR introduce a breaking change?** (check one)

- [x] No
- [ ] Yes (please describe)

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch, **not** the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (i.e. `[...] (fixes #xxx[,#xxx])`, where "xxx" is the issue number)
- [x] It should address only one issue or feature. If adding multiple features or fixing a bug and adding a new feature, break them into separate PRs if at all possible.
- [x] The title should follow the [**Conventional Commits**](https://www.conventionalcommits.org/) naming convention (i.e. `fix(alert): not alerting during SSR render`, `docs(badge): update pill examples, fix typos`, `chore: fix typo in README`, etc). **This is very important, as the `CHANGELOG` is generated from these messages.**

**If new features/enhancement/fixes are added or changed:**

- [x] Includes documentation updates (including updating the component's `package.json` for slot and event changes)
- [ ] New/updated tests are included and passing (if required)
- [x] Existing test suites are passing
- [x] The changes have not impacted the functionality of other components or directives
- [ ] ARIA Accessibility has been taken into consideration (Does it affect screen reader users or keyboard only users? Clickable items should be in the tab index, etc.)

**If adding a new feature, or changing the functionality of an existing feature, the PR's
description above includes:**

- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

### CODE REVIEW

File: src/components/index.esm.js [hunk 1]
```diff
@@ -41,11 +41,18 @@ export * from './pagination-nav'
 export * from './popover'
 export * from './progress'
 export * from './spinner'
-export * from './table'
+// Webpack 4 has difficulties with re-eport of re-export optimizations
```
Comment (jackmu95) [inline src/components/index.esm.js]: `re-export of re-exports`

File: src/components/index.esm.js [hunk 2]
```diff
@@ -41,11 +41,18 @@ export * from './pagination-nav'
...
+// So we import the table components indivisulaly here for better tree shaking.
+// the table components are one of hte largest components in the library
+// so we do this here to help with consumer optimizations.
+// Webpack v5 fixes the optimizations with re-export of re-exports so this
+// can be reverted back to `export * from './table'` when Webpack v5 is released.
+// https://github.com/webpack/webpack/pull/9203
+// export * from './table'
+export { default as BTable } from './table/table'
+export { default as BTableLite } from './table/table-lite'
```
Comment (jackmu95) [inline src/components/index.esm.js]: Maybe we should use this syntax for all components when webpack has difficulties with the other one.

File: src/components/index.esm.js (hunk 1 above)
Comment (tmorehouse) [reply src/components/index.esm.js]: Oops... was supposed to be commented out.

File: src/components/index.esm.js (hunk 2 above)
Comment (tmorehouse) [reply src/components/index.esm.js]: I thought about that...  but most component dirs have only one component (i.e. embed, jubotron, etc) or a bunch of smaller components (layout, dropdown) that would always be used together (or have insignificant size).

Then I thought of maintenance... we would need to maintain the imports (or automate the creation of the index file), when Webpack5 rolls out (they have https://www.npmjs.com/package/webpack/v/5.0.0-alpha.15 released at the moment) we would need to revert back.

What happens is that webpack will handle `export * from 'a'`, but doesn't properly track if `'a'` uses `export * from 'b'` (which is fixed in v5)

File: src/components/index.esm.js (hunk 1 above)
Comment (tmorehouse) [reply src/components/index.esm.js]: Yeah... structured exports exporting from the net level.

File: src/components/index.esm.js (hunk 2 above)
Comment (tmorehouse) [reply src/components/index.esm.js]: What webpack does (in some cases) is import  `src/components/table/index` into the final bundle, when all one imported was `BTableLite`.

The only other places where there would be a bit of savings would be with image (users may not use img-lazy)

And this is only applicable when importing individual components from a component group. Importing plugins has no issues.

Comment (tmorehouse) [reply src/components/index.esm.js]: Here is a breakdown of imported modules:

image

On the upper right is table (with the helper mixins and utils). it is bigger than modal.

Comment (tmorehouse) [reply src/components/index.esm.js]: But maybe we could do the re-export individually, until Webpack v5.0.0 is released.

Comment (tmorehouse) [reply src/components/index.esm.js]: I did find that

```js
export * from './table'
```
followed by
```js
export { default as BTable } from './table/table'
export { default as BTableLite } from './table/table-lite'
```

Worked well, and webpack and rollup correctly handled the duplicate export names

Comment (tmorehouse) [reply src/components/index.esm.js]: Although it really only benefits table/table-lite

Comment (tmorehouse) [reply src/components/index.esm.js]: But after a bit of thought, by exporting each component individually (for not util Webpack 5 or Webpack 4 is improved), does prevent the plugin code from being imported when not importing plugins.

**Pull request merged**